### PR TITLE
Add tie_weights() to LM heads and set bias in set_output_embeddings()

### PR DESCRIPTION
**Pull request by hackyon**
# What does this PR do?

This fixes a bug from the wrong bias in prediction heads in some situations. The `predictions.bias` needs to be tied to `predictions.decoder.bias` inside `tie_weights()`. 

Repro Steps:
1. Sync to HEAD in main
2. Add the following test case to `test_modeling_bert.py` and run it:
```
    def test_save_load_bert_prediction_head(self):
        with tempfile.TemporaryDirectory() as tmpdirname:
            model_to_save = BertForMaskedLM.from_pretrained("bert-base-uncased")

            model_to_save.save_pretrained(tmpdirname)

            model = BertForMaskedLM.from_pretrained(
                tmpdirname,
                low_cpu_mem_usage=True,
            )
            model.to(torch_device)
```
3. Error is thrown:
```
FAILED tests/models/bert/test_modeling_bert.py::BertModelTest::test_save_load_bert_prediction_head - NotImplementedError: Cannot copy out of meta tensor; no data!
```
```
    def convert(t):
        if convert_to_format is not None and t.dim() in (4, 5):
            return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None,
                        non_blocking, memory_format=convert_to_format)
>       return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
E       NotImplementedError: Cannot copy out of meta tensor; no data!

../venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1158: NotImplementedError
```

The issue was uncovered in https://github.com/huggingface/transformers/pull/28802.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @amyeroberts

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for digging into this and fixing for all these models! 

Just two things I think we need to do before merging: 
* Run the slow model tests for these models to confirm this change is backwards compatible.
* Apply the test to all models.

File: tests/models/deta/test_modeling_deta.py [hunk 1]
```diff
@@ -520,6 +520,10 @@ def test_initialization(self):
                         msg=f"Parameter {name} of model {model_class} seems not properly initialized",
                     )
 
+    @unittest.skip("Cannot be initialized on meta device as some weights are modified during the initialization")
```
Comment (hackyon) [inline tests/models/deta/test_modeling_deta.py:523]: This came from an existing comment:
https://github.com/huggingface/transformers/blob/354775bc5755c4a6c47e008d28f27f8ccdcf8f8f/src/transformers/models/deta/modeling_deta.py#L1780

It seems like the DetaForObjectDetection model has logic to manipulate the params during init, and so it simply doesn't support initializing on a meta device.

With that said, should we add some kind of flag in PreTrainedModel that would throw an error when trying to initial on meta device for models that do not support it? So far, it seems like DetaModel is the only one, so maybe not worth the effort to do so. 

cc @SunMarc who added this: https://github.com/huggingface/transformers/pull/27089

Comment (amyeroberts) [reply tests/models/deta/test_modeling_deta.py:523]: Let's keep as-is for the moment, until we have more affected models. DETR and DETR-like models are special but I think they're the only case.

**Review by amyeroberts — APPROVED** [review]
Great work - thanks for adding this!

**Pull request merged**